File: Cloude/Cloude/Models/Conversation.swift
import Foundation
import CloudeShared

struct Conversation: Codable, Identifiable {
    let id: UUID
    var name: String
    var symbol: String?
    var sessionId: String?
    var workingDirectory: String?
    let createdAt: Date
    var lastMessageAt: Date
    var messages: [ChatMessage]
    var pendingMessages: [ChatMessage]
    var pendingFork: Bool
    var defaultEffort: EffortLevel?
    var defaultModel: ModelSelection?
    var environmentId: UUID?
    var attachedBranch: String?
    var worktreePath: String?
    var originalWorkingDirectory: String?

    var isEmpty: Bool {
        messages.isEmpty && pendingMessages.isEmpty && sessionId == nil
    }

    var totalCost: Double {
        messages.compactMap(\.costUsd).reduce(0, +)
    }

    static let randomNames = [
        "Spark", "Nova", "Pulse", "Echo", "Drift", "Blaze", "Frost", "Dusk",
        "Dawn", "Flux", "Glow", "Haze", "Mist", "Peak", "Reef", "Sage",
        "Tide", "Vale", "Wave", "Zen", "Bolt", "Cove", "Edge", "Fern",
        "Grid", "Hive", "Jade", "Kite", "Leaf", "Maze", "Nest", "Opal",
        "Pine", "Quill", "Rush", "Sand", "Twig", "Vine", "Wisp", "Yarn",
        "Arc", "Bay", "Cliff", "Dell", "Elm", "Fog", "Glen", "Hill",
        "Ivy", "Jet", "Key", "Lane", "Moon", "Nook", "Oak", "Path"
    ]

    static let randomSymbols = [
        "star", "heart", "bolt", "flame", "leaf", "moon", "sun.max", "cloud",
        "sparkles", "wand.and.stars", "lightbulb", "paperplane", "rocket",
        "globe", "map", "flag", "bookmark", "tag", "bubble.left", "terminal",
        "paintbrush", "pencil", "folder", "doc", "book", "briefcase",
        "hammer", "wrench", "gearshape", "cpu", "lock", "key", "eye",
        "hare", "tortoise", "bird", "fish", "tree", "mountain.2", "drop"
    ]

    init(name: String? = nil, symbol: String? = nil, id: UUID = UUID(), sessionId: String? = nil, workingDirectory: String? = nil, pendingFork: Bool = false, environmentId: UUID? = nil) {
        self.id = id
        self.sessionId = sessionId
        self.workingDirectory = workingDirectory
        self.createdAt = Date()
        self.lastMessageAt = Date()
        self.messages = []
        self.pendingMessages = []
        self.pendingFork = pendingFork
        self.defaultEffort = nil
        self.defaultModel = nil
        self.environmentId = environmentId
        self.name = name ?? Self.randomNames.randomElement() ?? "Chat"
        self.symbol = symbol ?? Self.randomSymbols.randomElement()
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(UUID.self, forKey: .id)
        name = try container.decode(String.self, forKey: .name)
        symbol = try container.decodeIfPresent(String.self, forKey: .symbol)
        sessionId = try container.decodeIfPresent(String.self, forKey: .sessionId)
        workingDirectory = try container.decodeIfPresent(String.self, forKey: .workingDirectory)
        createdAt = try container.decode(Date.self, forKey: .createdAt)
        lastMessageAt = try container.decode(Date.self, forKey: .lastMessageAt)
        messages = try container.decode([ChatMessage].self, forKey: .messages)
        pendingMessages = try container.decodeIfPresent([ChatMessage].self, forKey: .pendingMessages) ?? []
        pendingFork = try container.decodeIfPresent(Bool.self, forKey: .pendingFork) ?? false
        defaultEffort = try container.decodeIfPresent(EffortLevel.self, forKey: .defaultEffort)
        defaultModel = try container.decodeIfPresent(ModelSelection.self, forKey: .defaultModel)
        environmentId = try container.decodeIfPresent(UUID.self, forKey: .environmentId)
        attachedBranch = try container.decodeIfPresent(String.self, forKey: .attachedBranch)
        worktreePath = try container.decodeIfPresent(String.self, forKey: .worktreePath)
        originalWorkingDirectory = try container.decodeIfPresent(String.self, forKey: .originalWorkingDirectory)
    }

    private enum CodingKeys: String, CodingKey {
        case id, name, symbol, sessionId, workingDirectory, createdAt, lastMessageAt, messages, pendingMessages, pendingFork, defaultEffort, defaultModel, environmentId, attachedBranch, worktreePath, originalWorkingDirectory
    }
}

File: android/app/src/main/java/com/cloude/app/Models/Conversation.kt
package com.cloude.app.Models

import kotlinx.serialization.Serializable
import java.util.UUID

@Serializable
data class Conversation(
    val id: String = UUID.randomUUID().toString(),
    var name: String = randomNames.random(),
    var symbol: String? = null,
    var sessionId: String? = null,
    var workingDirectory: String? = null,
    val createdAt: Long = System.currentTimeMillis(),
    var lastMessageAt: Long = System.currentTimeMillis(),
    var messages: MutableList<ChatMessage> = mutableListOf(),
    var pendingMessages: MutableList<ChatMessage> = mutableListOf(),
    var defaultEffort: String? = null,
    var defaultModel: String? = null,
    var environmentId: String? = null,
    var userRenamed: Boolean = false,
    var pendingFork: Boolean = false
) {
    val isEmpty: Boolean
        get() = messages.isEmpty() && pendingMessages.isEmpty() && sessionId == null

    val totalCost: Double
        get() = messages.mapNotNull { it.costUsd }.sum()

    companion object {
        val randomNames = listOf(
            "Spark", "Nova", "Pulse", "Echo", "Drift", "Blaze", "Frost", "Dusk",
            "Dawn", "Flux", "Glow", "Haze", "Mist", "Peak", "Reef", "Sage",
            "Tide", "Vale", "Wave", "Zen", "Bolt", "Cove", "Edge", "Fern",
            "Grid", "Hive", "Jade", "Kite", "Leaf", "Maze", "Nest", "Opal",
            "Pine", "Quill", "Rush", "Sand", "Twig", "Vine", "Wisp", "Yarn"
        )
    }
}

@Serializable
data class ChatMessage(
    val id: String = UUID.randomUUID().toString(),
    val isUser: Boolean,
    var text: String,
    val timestamp: Long = System.currentTimeMillis(),
    var toolCalls: MutableList<ToolCall> = mutableListOf(),
    var durationMs: Int? = null,
    var costUsd: Double? = null,
    var isQueued: Boolean = false,
    var wasInterrupted: Boolean = false,
    var serverUUID: String? = null,
    var model: String? = null,
    var isCollapsed: Boolean = false,
    var imageCount: Int = 0,
    var fileCount: Int = 0
)

@Serializable
data class ToolCall(
    val name: String,
    val input: String? = null,
    val toolId: String = UUID.randomUUID().toString(),
    val parentToolId: String? = null,
    var textPosition: Int? = null,
    var state: ToolCallState = ToolCallState.complete,
    var resultSummary: String? = null,
    var resultOutput: String? = null,
    var editInfo: EditInfo? = null
)

@Serializable
enum class ToolCallState {
    executing, complete
}

@Serializable
enum class EffortLevel(val displayName: String) {
    low("Low"), medium("Medium"), high("High"), max("Max")
}

@Serializable
enum class ModelSelection(val displayName: String) {
    opus("Opus"), sonnet("Sonnet"), haiku("Haiku")
}
